### PR TITLE
Add horizontal scrollbars and hide scrollbars if there is enough space

### PR DESCRIPTION
**Pull request by kxxt**
Aside from the title, this PR also contains some code for handling other scrollbar locations, but it's not exposed to user right now.

### CODE REVIEW

File: src/scroll_view.rs [hunk 1]
```diff
@@ -214,22 +304,118 @@ mod tests {
...
+        let (mut buf, scroll_buffer) = init_with_buffer_size(Size::new(25, 10));
+        let mut scroll_view_state = ScrollViewState::new();
+        scroll_buffer.render(Rect::new(2, 2, 20, 6), &mut buf, &mut scroll_view_state);
+        assert_buffer_eq!(
+            buf,
+            Buffer::with_lines(vec![
+                "                         ",
+                "                         ",
+                "  ABCDEFGHIJKLMNOPQRS▲   ",
+                "  UVWXYZABCDEFGHIJKLM█   ",
```
Comment (joshka) [inline src/scroll_view.rs:337]: This looks like an error - missing `T`

File: src/scroll_view.rs [hunk 2]
```diff
@@ -214,22 +304,118 @@ mod tests {
...
+                "  STUVWX  ",
+                "  MNOPQR  ",
+                "  GHIJKL  ",
+                "  ABCDEF  ",
+                "  UVWXYZ  ",
+                "  OPQRST  ",
+                "  IJKLMN  ",
+                "  CDEFGH  ",
+                "  WXYZAB  ",
+                "  ◄█═══►  ",
```
Comment (joshka) [inline src/scroll_view.rs:377]: Same here - looks like the last row is truncated.

**Review by joshka — COMMENTED** [review]
Looking good so far. There's a small off by 1 error I think, and some opportunity to simplify.

File: examples/scrollview.rs [hunk 3]
```diff
@@ -118,9 +120,9 @@ impl App {
     fn render_widgets_into_scrollview(&self, buf: &mut Buffer) {
         use Constraint::*;
         let area = buf.area;
-        // the scrollview (currently) allocates the full width of the buffer, but overwrites the
-        // last column with a scrollbar. This means that we need to account for this when laying
-        // out the widgets
+        // The scrollview (currently) allocates the full width of the buffer,
+        // but last row and column might be used by scrollbars.
+        // This means that we need to account for this when laying out the widgets
```
[the remark this answers is not in the document]
Comment (EdJoPaTo) [reply examples/scrollview.rs:125]: ratatui uses `rustfmt.toml` and `cargo +nightly fmt` for this.

[the remark this answers is not in the document]
Comment (kxxt) [reply examples/scrollview.rs]: Fixed

File: src/scroll_view.rs (hunk 1 above)
Comment (kxxt) [reply src/scroll_view.rs:337]: > This looks like an error - missing `T`

If I understand it correctly, the width of the view is 20 and the scrollbar takes the last col, so there are 19 cols left for the characters. On the first line, `T` is hidden by the scrollbar.

File: src/scroll_view.rs [hunk 4]
```diff
@@ -116,13 +116,33 @@ impl StatefulWidget for ScrollView {
...
+        let vertical = if self.size.height <= area.height {
+            // no vertical scrollbar, reset the y offset
+            y = 0;
+            None
+        } else {
+            Some(ScrollbarOrientation::VerticalRight)
+        };
+
+        let scrollbar_size = self.render_scrollbars(area, buf, state, horizontal, vertical);
         state.offset = (x, y).into();
```
[the remark this answers is not in the document]
Comment (kxxt) [reply src/scroll_view.rs]: I moved the decision about whether to show the scroll bar into the `render_scrollbars` func, while leaving the state calculation untouched.

File: src/scroll_view.rs [hunk 5]
```diff
@@ -136,11 +156,77 @@ impl ScrollView {
...
+    /// and return the size taken by the scrollbars
+    fn render_scrollbars(
+        &self,
+        area: Rect,
+        buf: &mut Buffer,
+        state: &ScrollViewState,
+        horizontal: Option<ScrollbarOrientation>,
+        vertical: Option<ScrollbarOrientation>,
+    ) -> Size {
+        let (pl, pt, pr, pb) = match (&horizontal, &vertical) {
```
[the remark this answers is not in the document]
Comment (kxxt) [reply src/scroll_view.rs]: > How about keeping this simple, and only supporting right and bottom for now, but adding the left/top as a future PR?

Done.

File: src/scroll_view.rs (hunk 1 above)
Comment (kxxt) [reply src/scroll_view.rs:310]: Oh, wait, this test hides horizontal scrollbar so it's indeed an off by one error

File: src/scroll_view.rs (hunk 2 above)
Comment (kxxt) [reply src/scroll_view.rs]: Fixed

File: examples/scrollview.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshka) [reply examples/scrollview.rs]: > ratatui uses `rustfmt.toml` and `cargo +nightly fmt` for this.

I use `cargo fmt -- --check` for all my personal repos. via https://github.com/joshka/github-workflows/blob/main/.github/workflows/rust-check.yml#L32. Neither approach catches manually wrapped shorter lines like this. I'm assuming that's to allow for reasonable situations where shorter lines make sense.

**Comment by joshka** [comment]
I just pushed a change which simplifies this a bunch (I hope) and catches an edge case - If there is enough vertical height to render the scrollview, and adding a horizontal scrollbar reduces the vertical space by one it's necessary to render a vertical scroll bar (and vice versa).

**Pull request merged**